node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,452 @@
+				let supermajority = supermajority_threshold(d.validators_for.len());
+				if d.validators_for.count_ones() >= supermajority ||
+					d.validators_against.count_ones() >= supermajority
+				{
+					partitioned.can_conclude_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.cant_conclude_onchain.push((session_index, candidate_hash));
+				}
+			},
+			None => partitioned.unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// TODO: Prioritize the votes in each partition. Is it still necessary?
+
+	// ... and CONCLUDED in two more
+	for (session_index, candidate_hash, _) in concluded {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(_) => partitioned.concluded_known_onchain.push((session_index, candidate_hash)),
+			None => partitioned.concluded_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn is_vote_worth_to_keep` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn vote_value() -> bool;
+}
+
+impl VoteType for InvalidDisputeStatementKind {
+	fn vote_value() -> bool {
+		false
+	}
+}
+
+impl VoteType for ValidDisputeStatementKind {
+	fn vote_value() -> bool {
+		true
+	}
+}
+
+/// Determines if a vote is worth to be kept, based on the onchain disputes
+fn is_vote_worth_to_keep<T: VoteType>(
+	v: &(T, ValidatorIndex, ValidatorSignature),
+	onchain_state: &DisputeState,
+) -> bool {
+	let local_vote = T::vote_value();
+	let (_, validator_index, _) = v;
+	let in_validators_for = onchain_state
+		.validators_for
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.copied()
+		.unwrap_or(false);
+	let in_validators_against = onchain_state
+		.validators_against
+		.get(validator_index.0 as usize)
+		.as_deref()
+		.copied()
+		.unwrap_or(false);
+
+	if in_validators_for && in_validators_against {
+		// The validator has double voted and runtime knows about this. Ignore this vote.
+		return false
+	}
+
+	if local_vote && in_validators_against || !local_vote && in_validators_for {
+		// local vote differs from the onchain vote
+		// we need this vote to punish the offending validator
+		return true
+	}
+
+	// The vote is valid. Return true if it is not seen onchain.
+	!in_validators_for && !in_validators_against
+}
+
+// Entrypoint to code relying on staging runtime API.
+// If the staging API is not enabled - returns an empty HashMap.
+async fn get_onchain_disputes<Sender>(
+	_sender: &mut Sender,
+	_relay_parent: Hash,
+) -> Result<HashMap<(SessionIndex, CandidateHash), DisputeState>, GetOnchainDisputesError>
+where
+	Sender: overseer::ProvisionerSenderTrait,
+{
+	let _onchain = Result::<
+		HashMap<(SessionIndex, CandidateHash), DisputeState>,
+		GetOnchainDisputesError,
+	>::Ok(HashMap::new());
+	#[cfg(feature = "staging-client")]
```

Comment:
With your upcoming staging API changes, we should differentiate here by version.